aws-cpp-sdk-core/source/http/crt/CRTHttpClient.cpp

```diff
@@ -0,0 +1,442 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/core/http/crt/CRTHttpClient.h>
+#include <aws/core/http/standard/StandardHttpResponse.h>
+#include <aws/core/http/standard/StandardHttpRequest.h>
+#include <aws/core/utils/ratelimiter/RateLimiterInterface.h>
+
+#include <aws/crt/http/HttpConnectionManager.h>
+#include <aws/crt/http/HttpRequestResponse.h>
+
+static const char *const CRT_HTTP_CLIENT_TAG = "CRTHttpClient";
+
+// Adapts AWS SDK input streams and rate limiters to the CRT input stream reading model.
+class ThrottleAwareInputStream : public Crt::Io::StdIOStreamInputStream {
+public:
+    ThrottleAwareInputStream(Utils::RateLimits::RateLimiterInterface& rateLimiter, std::shared_ptr<Aws::Crt::Io::IStream> stream,
+                             Aws::Crt::Allocator *allocator = Crt::ApiAllocator()) noexcept : Crt::Io::StdIOStreamInputStream(std::move(stream), allocator), m_rateLimiter(rateLimiter){
+
+    }
+protected:
+    bool ReadImpl(Crt::ByteBuf &buffer) noexcept override
+    {
+        // initial check to see if we should avoid reading for the moment.
+        if (m_rateLimiter.ApplyCost(0) == std::chrono::milliseconds(0)) {
+            size_t currentPos = buffer.len;
+
+            // now do the read. We may over read by an IO buffer size, but it's fine. The throttle will still
+            // kick-in in plenty of time.
+            bool retValue = Crt::Io::StdIOStreamInputStream::ReadImpl(buffer);
+
+            size_t newPos = buffer.len;
+            AWS_ASSERT(newPos >= currentPos && !"the buffer length should not have decreased in value.");
+            // now actually reduce the window.
+            m_rateLimiter.ApplyCost(static_cast<int64_t>(newPos - currentPos));
+            return retValue;
+        }
+
+        return true;
+    }
+
+private:
+    Utils::RateLimits::RateLimiterInterface& m_rateLimiter;
+};
+
+namespace Aws
+{
+    namespace Http
+    {
+        CRTHttpClient::CRTHttpClient(const Aws::Client::ClientConfiguration& clientConfig, Crt::Io::ClientBootstrap& bootstrap) : 
+            HttpClient(), m_context(), m_proxyOptions(), m_bootstrap(bootstrap), m_configuration(clientConfig)
+        {
+            //first need to figure TLS out...
+            Crt::Io::TlsContextOptions tlsContextOptions = Crt::Io::TlsContextOptions::InitDefaultClient();
+
+            CheckAndInitializeProxySettings(clientConfig);
+
+            // if ca is overridden and a proxy is configured, it's intended for the proxy, not this context.

```

Comment:
If we ever start supporting SOCKS proxy this statement will become invalid.

Reply:
Do we currently support SOCKS proxies?

Reply:
We are getting such [requests](https://github.com/aws/aws-sdk-cpp/pull/1562) from time to time, but it is still pending an internal review.

Reply:
updated the comment to make sure it was called out this is based on the current SDK configuration capabilities.